vllm/hpu/ops.py

```diff
@@ -25,72 +25,247 @@ def gelu_new(output, input):
+                                                )
+
+                                    with torch.profiler.record_function("store_seq_weights"):
+                                        attn_weights.index_copy_(0, seq_index, attn_weight.unsqueeze(0))
+
+                        with torch.profiler.record_function("seq_index_inc"):
+                            seq_index.add_(1)
+            else:
+                attn_weights = attn_weights_2
+
+            with torch.profiler.record_function("store_block_weight"):
+                attn_weights_blocks.index_copy_(0, block_index, attn_weights.unsqueeze(0))
+
+        with torch.profiler.record_function("block_index_inc"):
+            block_index.add_(1)
+        if device == "hpu":
+            htorch.core.mark_step()
+
+    # Cleanup out-of-bound weights and values
+    attn_weights_blocks_filler = torch.finfo(query.dtype).min
+    value_blocks_filler = 0.0
+    _paged_attention_masked_fill(
+        attn_weights_blocks,
+        value_blocks,
+        context_lens,
+        attn_weights_blocks_filler,
+        value_blocks_filler,
+        block_size,
+        max_num_blocks_per_seq,
+        num_seqs,
+        sanitize_values,
+        device,
+    )
+
+    exp_sum = attn_weights_blocks.exp().sum(dim=-1, keepdim=True).sum(0)
 
     output = torch.zeros_like(query)
-    for i in range(len(attn_weights_blocks)):
-        attention_probs = torch.exp(attn_weights_blocks[i]) / exp_sum
-        value = value_blocks[i]
-        out = torch.matmul(attention_probs.to(value.dtype), value).reshape(num_seqs, num_query_heads, head_size)
-        output.add_(out)
-    htorch.core.mark_step()
+    for i in range(max_num_blocks_per_seq):
+        with torch.profiler.record_function(f"output_block_loop_{i}"):
+            with torch.profiler.record_function("compute_block_softmax"):
+                attention_probs = torch.exp(attn_weights_blocks[i]) / exp_sum
+            value = value_blocks[i]
+            if num_queries_per_kv > 1:
+                with torch.profiler.record_function("repeat_interleave_block_value"):
+                    # Handle MQA and GQA
+                    value_4d_view = value.reshape(num_seqs, num_kv_heads, block_size, head_size)
+                    value = torch.repeat_interleave(value_4d_view, num_queries_per_kv, dim=1)
+            with torch.profiler.record_function("reshape_for_gemm"):
+                attention_probs = attention_probs.to(value.dtype).reshape(num_seqs * num_query_heads, 1, block_size)
+                value = value.reshape(num_seqs * num_query_heads, block_size, head_size)
+            with torch.profiler.record_function("gemm_out"):
+                out = torch.matmul(attention_probs, value).reshape(num_seqs, num_query_heads, head_size)
+            with torch.profiler.record_function("gemm_accumulate"):
+                output.add_(out)
+    if device == "hpu":
+        htorch.core.mark_step()
     return output.to(dtype=query_in.dtype)
```

Comment:
There's no type casting of the query in the current code, so this casting of the output is superfluous.